HTTP Methods Examples Provided 20231026 (#6756)
opensearch-project/documentation-website: _security/access-control/document-level-security.md

parent: Access control
nav_order: 85
redirect_from:
- /security/access-control/document-level-security/
- /security-plugin/access-control/document-level-security/
---

> # Document-level security (DLS)

Reviewer: ```suggestion
# Document-level security
```

> Document-level security (DLS) lets you restrict a role to a subset of documents in an index. The easiest way to get started with document- and field-level security is to open OpenSearch Dashboards and choose **Security**. Then choose **Roles**, create a new role, and review the **Index Permissions** section. See the Index Permissions section in OpenSearch Dashboards in the following image.

Reviewer: ```suggestion
Document-level security lets you restrict a role to a subset of documents in an index. The easiest way to get started with document- and field-level security is to open OpenSearch Dashboards and choose **Security**. Then choose **Roles**, create a new role, and review the **Index Permissions** section, shown in the following image.
```

> ![Document- and field-level security screen in OpenSearch Dashboards]({{site.url}}{{site.baseurl}}/images/security-dls.png)

Reviewer: Please reference in the preceding text.

## Simple roles

> Document-level security uses the OpenSearch query DSL to define which documents a role grants access to. In OpenSearch Dashboards, choose an index pattern and provide a query in the **Document level security** section:

Reviewer: ```suggestion
Document-level security uses OpenSearch query DSL to define which documents a role grants access to. In OpenSearch Dashboards, choose an index pattern and provide a query in the **Document-level security** section:
```

> Document-level security uses the OpenSearch query DSL to define which documents a role grants access to. In OpenSearch Dashboards, choose an index pattern and provide a query in the **Document level security** section:

Reviewer: DSL should be defined on first appearance.

> Document-level security uses OpenSearch query DSL to define which documents a role grants access to. In OpenSearch Dashboards, choose an index pattern and provide a query in the **Document-level security** section:

Reviewer: ```suggestion
Document-level security uses OpenSearch query domain-specific language (DSL) to define which documents a role grants access to. In OpenSearch Dashboards, choose an index pattern and provide a query in the **Document-level security** section:
```

```json
{
  "bool": {
    "must": {
      "match": {
        "genres": "Comedy"
      }
    }
  }
}
```

This query specifies that for the role to have access to a document, its `genres` field must include `Comedy`.

> A typical request to the `_search` API includes `{ "query": { ... } }` around the query, but in this case, you only need to specify the query itself.

Reviewer: ```suggestion
A typical request sent to the `_search` API includes `{ "query": { ... } }` around the query, but in this case, you only need to specify the query itself.
```

A typical request to the `_search` API includes `{ "query": { ... } }` around the query, but in this case, you only need to specify the query itself.

In the REST API, you provide the query as a string, so you must escape your quotes. This role allows a user to read any document in any index with the field `public` set to `true`:
---
layout: default
title: Document-level security
parent: Access control
nav_order: 85
redirect_from:
- /security/access-control/document-level-security/
- /security-plugin/access-control/document-level-security/
---

# Document-level security (DLS)

> Document-level security lets you restrict a role to a subset of documents in an index. The easiest way to get started with document- and field-level security is to open OpenSearch Dashboards and choose **Security**. Then choose **Roles**, create a new role, and review the **Index permissions** section.

Reviewer 2: Should "permissions" in index permissions be capitalized as a proper noun here?

> ![Document- and field-level security screen in OpenSearch Dashboards]({{site.url}}{{site.baseurl}}/images/security-dls.png)

Reviewer 2: @kolchfa-aws can you confirm if the hyphen syntax is correct here?

Author: It's a good observation, I was following the page as it was previously but it would be good to get another opinion on what is preferred please @kolchfa-aws

  ]
}
```

**Role B without DLS:** This role is specifically granted to certain users, such as administrators, and does not include DLS, as shown in the following permission set:

```
{
  "index_permissions" : [
    {
      "index_patterns" : ["*"],
      "allowed_actions" : ["indices:data/read/search"]
    }
  ]
}
```
{% include copy.html %}

Setting `plugins.security.dfm_empty_overrides_all` to `true` ensures that administrators assigned Role B can override any DLS restrictions imposed by Role A. This allows specific Role B users to access all documents, regardless of the restrictions applied by Role A's DLS restrictions.

#### Example: Search template access

In this example, two roles are defined, one with DLS and another without DLS, granting access to search templates:

**Role A with DLS:**

```
{
  "index_permissions": [
    {
      "index_patterns": [
        "example-index"
      ],
      "dls": "[.. some DLS here ..]",
      "allowed_actions": [
        "indices:data/read/search",
      ]
    }
  ]
}
```
{% include copy.html %}

**Role B, without DLS**, which only grants access to search templates:

```
{
  "index_permissions" : [
    {
      "index_patterns" : [ "*" ],

>       "allowed_actions" : [ "indices:data/read/search/template" ]---

Reviewer 2: This looks like the formatting is weird?

Reviewer 3: A copy and paste error maybe @leanneeliatra?

Author: Indeed, thanks

layout: default
title: Document-level security
parent: Access control
nav_order: 85
redirect_from:
- /security/access-control/document-level-security/
- /security-plugin/access-control/document-level-security/
---

# Document-level security (DLS)

Document-level security lets you restrict a role to a subset of documents in an index. The easiest way to get started with document- and field-level security is to open OpenSearch Dashboards and choose **Security**. Then choose **Roles**, create a new role, and review the **Index permissions** section.

![Document- and field-level security screen in OpenSearch Dashboards]({{site.url}}{{site.baseurl}}/images/security-dls.png)

## Simple roles

Document-level security uses the OpenSearch query DSL to define which documents a role grants access to. In OpenSearch Dashboards, choose an index pattern and provide a query in the **Document level security** section:

```json
{
  "bool": {
    "must": {
      "match": {
        "genres": "Comedy"
      }
    }
  }
}
```

> This query specifies that for the role to have access to a document, its `genres` field must include `Comedy`.

Reviewer 2: Is this supposed to be repeated?

Author: That was an error Stephen thanks for catching.
